// @reclaimprotocol/witness-sdk/lib/beacon/smart-contract/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const config_1 = require("../../config");
const utils_1 = require("../../utils");
const utils_2 = require("./utils");
function makeSmartContractBeacon(chainId) {
    chainId = chainId || config_1.DEFAULT_CHAIN_ID;
    const contract = (0, utils_2.getContract)(chainId);
    return (0, utils_1.makeBeaconCacheable)({
        async getState(epochId) {
            const epoch = await contract.fetchEpoch(epochId || 0);
            if (!epoch.id) {
                throw new Error(`Invalid epoch ID: ${epochId}`);
            }
            return {
                epoch: epoch.id,
                witnesses: epoch.witnesses.map(w => ({
                    id: w.addr.toLowerCase(),
                    url: w.host
                })),
                witnessesRequiredForClaim: epoch.minimumWitnessesForClaimCreation,
                nextEpochTimestampS: epoch.timestampEnd
            };
        }
    });
}
exports.default = makeSmartContractBeacon;
