// @reclaimprotocol/witness-sdk/lib/providers/facebook-friends-count.js
"use strict";
/**
 * Number of friends on facebook
 */
Object.defineProperty(exports, "__esModule", { value: true });
const config_1 = require("../config");
const http_parser_1 = require("../utils/http-parser");
const HOST = 'www.facebook.com';
const HOSTPORT = `${HOST}:${config_1.DEFAULT_PORT}`;
const facebookFriendsCount = {
    hostPort: HOSTPORT,
    areValidParams(params) {
        return (typeof params.userURL === 'string' &&
            params.userURL !== '' &&
            typeof params.friendsCount === 'number' &&
            params.friendsCount >= 0);
    },
    createRequest(secretParams, params) {
        const data = [
            `GET ${params.userURL} HTTP/1.1`,
            'Host: www.facebook.com',
            'authority: www.facebook.com',
            'Connection: close',
            'accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8',
            `cookie: ${secretParams.cookie};`,
            'Accept-Encoding: identity',
            'User-Agent: reclaim/1.0.0',
            '\r\n',
        ].join('\r\n');
        // Find the cookie and redact it
        const cookieStartIndex = data.indexOf(secretParams.cookie);
        return {
            data,
            redactions: [
                {
                    fromIndex: cookieStartIndex,
                    toIndex: cookieStartIndex + secretParams.cookie.length,
                },
            ],
        };
    },
    assertValidProviderReceipt(receipt, params) {
        if (receipt.hostPort !== HOSTPORT) {
            throw new Error(`Invalid hostPort: ${receipt.hostPort}`);
        }
        const req = (0, http_parser_1.getHttpRequestHeadersFromTranscript)(receipt.transcript);
        if (req.method !== 'get') {
            throw new Error(`Invalid method: ${req.method}`);
        }
        if (!req.url.startsWith(params.userURL)) {
            throw new Error(`Invalid URL: ${req.url}`);
        }
        if (req.headers['connection'] !== 'close') {
            throw new Error('Invalid connection header');
        }
        // Parse the HTTP response
        const res = (0, http_parser_1.getCompleteHttpResponseFromTranscript)(receipt.transcript);
        if (res.statusCode !== 200) {
            throw new Error(`Invalid status code: ${res.statusCode}`);
        }
        const bodyStr = res.body.toString();
        const friendsCountRegex = /\d+ friends/g;
        const friends = bodyStr.match(friendsCountRegex);
        let resFriendsCount = 0;
        if (friends) {
            resFriendsCount = parseInt(friends[0].split(' ')[0]);
        }
        if (resFriendsCount !== params.friendsCount) {
            throw new Error('Friends count not equal to user specified value');
        }
    },
};
exports.default = facebookFriendsCount;
