// @reclaimprotocol/witness-sdk/lib/providers/godaddy-user.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Godaddy Professional Email
 *
 * https://eu1.myprofessionalmail.com/appsuite/api/account
 */
const config_1 = require("../config");
const utils_1 = require("../utils");
const http_parser_1 = require("../utils/http-parser");
// where to send the HTTP request
const HOST = (region) => `${region}.myprofessionalmail.com`;
const HOSTPORT = (params) => `${HOST(params.region)}:${config_1.DEFAULT_PORT}`;
// what API to call
const METHOD = 'GET';
const URL = '/appsuite/api/account';
const goDaddyLogin = {
    hostPort: HOSTPORT,
    areValidParams(params) {
        return typeof params.emailAddress === 'string' && typeof params.region === 'string';
    },
    createRequest({ cookieStr, encQueryParams }, { region }) {
        // serialise the HTTP request
        const endpoint = URL + '?' + encQueryParams;
        const data = [
            `${METHOD} ${endpoint} HTTP/1.1`,
            'Host: ' + HOST(region),
            'cookie: ' + cookieStr,
            'Connection: close',
            'Content-Length: 0',
            '\r\n'
        ].join('\r\n');
        // find the token and redact it
        const tokenStartIndex = data.indexOf(cookieStr);
        const encParamsStartIndex = data.indexOf(encQueryParams);
        return {
            data,
            // anything that should be redacted from the transcript
            // should be added to this array
            redactions: [
                {
                    fromIndex: tokenStartIndex,
                    toIndex: tokenStartIndex + cookieStr.length,
                },
                {
                    fromIndex: encParamsStartIndex,
                    toIndex: encParamsStartIndex + encQueryParams.length,
                }
            ]
        };
    },
    assertValidProviderReceipt(receipt, params) {
        var _a, _b, _c;
        // ensure the request was sent to the right place
        if (receipt.hostPort !== HOSTPORT(params)) {
            throw new Error(`Invalid hostPort: ${receipt.hostPort}`);
        }
        // parse the HTTP request & check
        // the method, URL, headers, etc. match what we expect
        const req = (0, http_parser_1.getHttpRequestHeadersFromTranscript)(receipt.transcript);
        if (req.method !== METHOD.toLowerCase()) {
            throw new Error(`Invalid method: ${req.method}`);
        }
        if (!req.url.startsWith(URL)) {
            throw new Error(`Invalid URL: ${req.url}`);
        }
        // we ensure the connection header was sent as "close"
        // this is done to avoid any possible malicious request
        // that contains multiple requests, but via redactions
        // is spoofed as a single request
        if (req.headers['connection'] !== 'close') {
            throw new Error('Invalid connection header');
        }
        // now we parse the HTTP response & check
        // if the emailAddress returned by the API
        // matches the parameters the user provided
        const res = (0, http_parser_1.getCompleteHttpResponseFromTranscript)(receipt.transcript);
        if (res.statusCode !== 200) {
            throw new Error(`Invalid status code: ${res.statusCode}`);
        }
        if (!((_a = res.headers['content-type']) === null || _a === void 0 ? void 0 : _a.startsWith('application/json'))) {
            throw new Error(`Invalid content-type: ${res.headers['content-type']}`);
        }
        // parse the body & try to find the email add
        // that the user provided in the parameters
        const json = JSON.parse((0, utils_1.uint8ArrayToStr)(res.body));
        const resAddr = (_c = (_b = json === null || json === void 0 ? void 0 : json.data) === null || _b === void 0 ? void 0 : _b[0]) === null || _c === void 0 ? void 0 : _c[0];
        if (resAddr !== params.emailAddress) {
            throw new Error(`Email "${params.emailAddress}" could not be matched`);
        }
    },
};
exports.default = goDaddyLogin;
