// @reclaimprotocol/witness-sdk/lib/providers/http-provider/wrapper.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.wrapInHttpProvider = void 0;
const _1 = __importDefault(require("."));
/**
 * Wrap a provider in an HTTP provider
 */
function wrapInHttpProvider({ getParams, getSecretParams, areValidParams, }) {
    return {
        hostPort(params) {
            if (typeof _1.default.hostPort === 'string') {
                return _1.default.hostPort;
            }
            return _1.default.hostPort(getParams(params));
        },
        areValidParams: areValidParams,
        createRequest(secretParams, params) {
            return _1.default.createRequest(getSecretParams(secretParams), getParams(params));
        },
        assertValidProviderReceipt(receipt, params) {
            return _1.default.assertValidProviderReceipt(receipt, getParams(params));
        },
        getResponseRedactions(response, params) {
            return _1.default.getResponseRedactions(response, getParams(params));
        },
    };
}
exports.wrapInHttpProvider = wrapInHttpProvider;
