// @reclaimprotocol/witness-sdk/lib/providers/spotify/spotify-email.js
"use strict";
/**
 * Verify spotify user email
 *
 * https://www.spotify.com/api/account-settings/v1/profile
 */
Object.defineProperty(exports, "__esModule", { value: true });
const config_1 = require("../../config");
const utils_1 = require("../../utils");
const http_parser_1 = require("../../utils/http-parser");
// where to send the HTTP request
const HOST = 'www.spotify.com';
const HOSTPORT = `${HOST}:${config_1.DEFAULT_PORT}`;
// what API to call
const METHOD = 'GET';
const URL = '/api/account-settings/v1/profile';
const spotifyEmail = {
    hostPort: HOSTPORT,
    areValidParams(params) {
        return typeof (params === null || params === void 0 ? void 0 : params.userEmail) === 'string';
    },
    createRequest({ token }) {
        // serialise the HTTP request
        const url = URL;
        const data = [
            `${METHOD} ${url} HTTP/1.1`,
            'Host: ' + HOST,
            `cookie: sp_dc=${token};`,
            'accept: */*',
            'user-agent: reclaim/1.0.0',
            'Content-Length: 0',
            'Connection: close',
            '\r\n'
        ].join('\r\n');
        // find the token and redact it
        const tokenStartIndex = data.indexOf(token);
        return {
            data,
            // anything that should be redacted from the transcript
            // should be added to this array
            redactions: [
                {
                    fromIndex: tokenStartIndex,
                    toIndex: tokenStartIndex + token.length
                }
            ]
        };
    },
    assertValidProviderReceipt(receipt, { userEmail }) {
        var _a;
        // ensure the request was sent to the right place
        if (receipt.hostPort !== HOSTPORT) {
            throw new Error(`Invalid hostPort: ${receipt.hostPort}`);
        }
        // parse the HTTP request & check
        // the method, URL, headers, etc. match what we expect
        const req = (0, http_parser_1.getHttpRequestHeadersFromTranscript)(receipt.transcript);
        if (req.method !== METHOD.toLowerCase()) {
            throw new Error(`Invalid method: ${req.method}`);
        }
        if (!req.url.startsWith(URL)) {
            throw new Error(`Invalid URL: ${req.url}`);
        }
        // we ensure the connection header was sent as "close"
        // this is done to avoid any possible malicious request
        // that contains multiple requests, but via redactions
        // is spoofed as a single request
        if (req.headers['connection'] !== 'close') {
            throw new Error('Invalid connection header');
        }
        // now we parse the HTTP response
        const res = (0, http_parser_1.getCompleteHttpResponseFromTranscript)(receipt.transcript);
        if (res.statusCode !== 200) {
            throw new Error(`Invalid status code: ${res.statusCode}`);
        }
        const json = JSON.parse((0, utils_1.uint8ArrayToStr)(res.body));
        const resEmail = (_a = json.profile) === null || _a === void 0 ? void 0 : _a.email;
        if (resEmail !== userEmail) {
            throw new Error('Email does not match');
        }
    },
};
exports.default = spotifyEmail;
