// @reclaimprotocol/witness-sdk/lib/providers/swiggy-min-order/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const config_1 = require("../../config");
const utils_1 = require("../../utils");
const http_parser_1 = require("../../utils/http-parser");
const HOST = 'www.swiggy.com';
const HOSTPORT = `${HOST}:${config_1.DEFAULT_PORT}`;
const PATH = '/dapi/order/all?order_id=';
const SwiggyTotalOrder = {
    hostPort: HOSTPORT,
    areValidParams(params) {
        return typeof params.orderCount === 'number' && params.orderCount >= 0;
    },
    createRequest({ cookieStr }) {
        const data = [
            `GET ${PATH} HTTP/1.1`,
            'Host: ' + HOST,
            'Connection: close',
            `cookie: ${cookieStr}`,
            'User-Agent: reclaim/1.0.0',
            'Accept-Encoding: gzip, deflate',
            '\r\n',
        ].join('\r\n');
        const cookieStartIndex = data.indexOf(cookieStr);
        return {
            data,
            redactions: [
                {
                    fromIndex: cookieStartIndex,
                    toIndex: cookieStartIndex + cookieStr.length,
                },
            ],
        };
    },
    assertValidProviderReceipt(receipt, { orderCount }) {
        var _a, _b, _c, _d, _e;
        if (receipt.hostPort !== HOSTPORT) {
            throw new Error(`Invalid hostPort: ${receipt.hostPort}`);
        }
        if (receipt.hostPort !== HOSTPORT) {
            throw new Error(`Invalid hostPort: ${receipt.hostPort}`);
        }
        const req = (0, http_parser_1.getHttpRequestHeadersFromTranscript)(receipt.transcript);
        if (req.method !== 'get') {
            throw new Error(`Invalid method: ${req.method}`);
        }
        const res = (0, http_parser_1.getCompleteHttpResponseFromTranscript)(receipt.transcript);
        if (!((_a = res.headers['content-type']) === null || _a === void 0 ? void 0 : _a.startsWith('application/json'))) {
            throw new Error(`Invalid content-type: ${res.headers['content-type']}`);
        }
        let html;
        if (res.headers['content-encoding'] === 'gzip') {
            const buf = Buffer.from(res.body);
            html = (0, utils_1.gunzipSync)(buf).toString();
        }
        else {
            html = res.body.toString();
        }
        const data = JSON.parse(html);
        if (((_c = (_b = data === null || data === void 0 ? void 0 : data.data) === null || _b === void 0 ? void 0 : _b.orders) === null || _c === void 0 ? void 0 : _c.length) === 'undefined') {
            throw new Error("Can't fetch order count");
        }
        const resOrderCount = (_e = (_d = data === null || data === void 0 ? void 0 : data.data) === null || _d === void 0 ? void 0 : _d.orders) === null || _e === void 0 ? void 0 : _e.length;
        if (resOrderCount !== orderCount) {
            throw new Error("OrderCount Doesn't match");
        }
    },
};
exports.default = SwiggyTotalOrder;
