// @reclaimprotocol/witness-sdk/lib/providers/tinder-match-count/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const config_1 = require("../../config");
const utils_1 = require("../../utils");
const http_parser_1 = require("../../utils/http-parser");
// where to send the HTTP request
const HOST = 'api.gotinder.com';
const HOSTPORT = `${HOST}:${config_1.DEFAULT_PORT}`;
const PATH = '/v2/matches?locale=en-GB&count=60&is_tinder_u=false';
const TinderMatchCount = {
    hostPort: HOSTPORT,
    areValidParams(params) {
        return (typeof params.userId === 'string' &&
            params.userId !== '' &&
            typeof params.matchCount === 'number' &&
            params.matchCount >= 0);
    },
    createRequest({ token }) {
        const data = [
            `GET ${PATH} HTTP/1.1`,
            'Host: ' + HOST,
            `X-Auth-Token: ${token}`,
            'Connection: close',
            'User-Agent: reclaim/1.0.0',
            'Accept-Encoding: gzip, deflate',
            '\r\n',
        ].join('\r\n');
        const tokenStartIndex = data.indexOf(token);
        return {
            data,
            redactions: [
                {
                    fromIndex: tokenStartIndex,
                    toIndex: tokenStartIndex + token.length,
                },
            ],
        };
    },
    assertValidProviderReceipt(receipt, { matchCount }) {
        var _a, _b, _c, _d;
        // ensure the request was sent to the right place
        if (receipt.hostPort !== HOSTPORT) {
            throw new Error(`Invalid hostPort: ${receipt.hostPort}`);
        }
        if (receipt.hostPort !== HOSTPORT) {
            throw new Error(`Invalid hostPort: ${receipt.hostPort}`);
        }
        // parse the HTTP request & check
        // the method, URL, headers, etc. match what we expect
        const req = (0, http_parser_1.getHttpRequestHeadersFromTranscript)(receipt.transcript);
        if (req.method !== 'get') {
            throw new Error(`Invalid method: ${req.method}`);
        }
        const res = (0, http_parser_1.getCompleteHttpResponseFromTranscript)(receipt.transcript);
        if (!((_a = res.headers['content-type']) === null || _a === void 0 ? void 0 : _a.startsWith('application/json'))) {
            throw new Error(`Invalid content-type: ${res.headers['content-type']}`);
        }
        let html;
        if (res.headers['content-encoding'] === 'gzip') {
            const buf = Buffer.from(res.body);
            html = (0, utils_1.gunzipSync)(buf).toString();
        }
        else {
            html = res.body.toString();
        }
        const data = JSON.parse(html);
        const resMatchCount = (_d = (_c = (_b = data === null || data === void 0 ? void 0 : data.data) === null || _b === void 0 ? void 0 : _b.matches) === null || _c === void 0 ? void 0 : _c.length) !== null && _d !== void 0 ? _d : 0;
        if (resMatchCount < matchCount) {
            throw new Error('Match count less than user specified value');
        }
    },
};
exports.default = TinderMatchCount;
